File: src/orchestrator/compaction.py
"""Orchestrator-level compaction and state checkpoint/restore.

Manages the orchestrator's own context window by periodically saving
and restoring state snapshots.  Also provides threshold-based compaction
decisions for individual agents.
"""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta

from sqlalchemy import func, select, update

from orchestrator.memory import MemoryManager
from orchestrator.models import Agent, Memory, Task


@dataclass
class OrchestratorSnapshot:
    """State snapshot used to resume the orchestrator after a restart."""

    active_tasks: list[Task] = field(default_factory=list)
    pending_reviews: list[Task] = field(default_factory=list)
    active_agents: list[Agent] = field(default_factory=list)
    stale_agents: list[Agent] = field(default_factory=list)
    last_checkpoint: datetime | None = None


class OrchestratorCompaction:
    """Manages orchestrator state checkpoints and compaction decisions."""

    # Heartbeats older than this are considered stale.
    STALE_HEARTBEAT_SECONDS = 120

    def __init__(
        self,
        memory_manager: MemoryManager,
        compaction_threshold: float = 0.7,
    ) -> None:
        self._memory = memory_manager
        self._compaction_threshold = compaction_threshold

    # ------------------------------------------------------------------
    # Save / Restore
    # ------------------------------------------------------------------

    async def save_orchestrator_state(
        self, orchestrator_agent_id: uuid.UUID
    ) -> None:
        """Periodic checkpoint of orchestrator state.

        Saves:
        - Current task assignments (which agent has which task)
        - Pending decisions and their context
        - Recent events summary
        - Known blockers and their status

        Stored as Memory records with type ``orchestrator_state``.
        """
        async with self._memory._session_factory() as session:
            # Gather task assignments
            stmt = (
                select(Task)
                .where(Task.status == "in_progress")
                .where(Task.assigned_agent_id.isnot(None))
            )
            result = await session.execute(stmt)
            in_progress = list(result.scalars().all())

            assignments = [
                {
                    "task_id": str(t.id),
                    "title": t.title,
                    "agent_id": str(t.assigned_agent_id),
                    "status": t.status,
                }
                for t in in_progress
            ]

            # Pending reviews
            stmt_review = select(Task).where(
                Task.status.in_(["plan_review", "testing_ready"])
            )
            result_review = await session.execute(stmt_review)
            reviews = list(result_review.scalars().all())

            pending = [
                {
                    "task_id": str(t.id),
                    "title": t.title,
                    "status": t.status,
                }
                for t in reviews
            ]

            # Active agents
            stmt_agents = select(Agent).where(Agent.status == "working")
            result_agents = await session.execute(stmt_agents)
            agents = list(result_agents.scalars().all())

            active_agents = [
                {
                    "agent_id": str(a.id),
                    "name": a.name,
                    "current_task_id": (
                        str(a.current_task_id)
                        if a.current_task_id
                        else None
                    ),
                    "heartbeat_at": (
                        a.heartbeat_at.isoformat()
                        if a.heartbeat_at
                        else None
                    ),
                }
                for a in agents
            ]

            # Known blockers
            blocker_stmt = (
                select(Memory)
                .where(Memory.agent_id == orchestrator_agent_id)
                .where(Memory.memory_type == "blocker")
                .order_by(Memory.created_at.desc())
                .limit(20)
            )
            blocker_result = await session.execute(blocker_stmt)
            blockers = [m.content for m in blocker_result.scalars().all()]

            # Build state content
            state = {
                "assignments": assignments,
                "pending_reviews": pending,
                "active_agents": active_agents,
                "blockers": blockers,
                "checkpoint_at": datetime.now(UTC).isoformat(),
            }

        # Store as a memory record
        await self._memory.store_memory(
            agent_id=orchestrator_agent_id,
            content=json.dumps(state),
            memory_type="orchestrator_state",
            metadata={"version": 1},
        )

    async def restore_orchestrator_state(
        self, orchestrator_agent_id: uuid.UUID
    ) -> OrchestratorSnapshot:
        """Restore orchestrator state after restart.

        1. Load latest ``orchestrator_state`` memory.
        2. Load all in-progress tasks.
        3. Load all active agents and their heartbeats.
        4. Reconcile: mark agents with stale heartbeats as DEAD.
        5. Return snapshot for the orchestrator to resume from.
        """
        snapshot = OrchestratorSnapshot()

        async with self._memory._session_factory() as session:
            # 1. Latest orchestrator_state checkpoint
            state_stmt = (
                select(Memory)
                .where(Memory.agent_id == orchestrator_agent_id)
                .where(Memory.memory_type == "orchestrator_state")
                .order_by(Memory.created_at.desc())
                .limit(1)
            )
            state_result = await session.execute(state_stmt)
            state_mem = state_result.scalar_one_or_none()
            if state_mem:
                snapshot.last_checkpoint = state_mem.created_at

            # 2. In-progress tasks
            task_stmt = select(Task).where(Task.status == "in_progress")
            task_result = await session.execute(task_stmt)
            snapshot.active_tasks = list(task_result.scalars().all())

            # Pending reviews
            review_stmt = select(Task).where(
                Task.status.in_(["plan_review", "testing_ready"])
            )
            review_result = await session.execute(review_stmt)
            snapshot.pending_reviews = list(review_result.scalars().all())

            # 3. Working agents
            agent_stmt = select(Agent).where(Agent.status == "working")
            agent_result = await session.execute(agent_stmt)
            working_agents = list(agent_result.scalars().all())

            # 4. Reconcile stale heartbeats
            now = datetime.now(UTC)
            cutoff = now - timedelta(seconds=self.STALE_HEARTBEAT_SECONDS)

            for agent in working_agents:
                heartbeat = agent.heartbeat_at
                # Normalise naive datetimes (e.g. from SQLite) to UTC
                if heartbeat is not None and heartbeat.tzinfo is None:
                    heartbeat = heartbeat.replace(tzinfo=UTC)
                if heartbeat is None or heartbeat < cutoff:
                    # Mark as dead
                    await session.execute(
                        update(Agent)
                        .where(Agent.id == agent.id)
                        .values(status="dead")
                    )
                    agent.status = "dead"
                    snapshot.stale_agents.append(agent)
                else:
                    snapshot.active_agents.append(agent)

            await session.commit()

        return snapshot

    # ------------------------------------------------------------------
    # Compaction decision
    # ------------------------------------------------------------------

    async def should_compact(self, agent_id: uuid.UUID) -> bool:
        """Determine if an agent's memory needs compaction.

        Based on total memory count and estimated token size.
        Returns True if over threshold.

        The threshold is expressed as a fraction (0..1).  We consider
        compaction necessary when:
        - memory count > 50 * threshold  (i.e. ~35 at 0.7), OR
        - estimated tokens > 8000 * threshold  (i.e. ~5600 at 0.7)
        """
        async with self._memory._session_factory() as session:
            # Count non-archived memories
            count_stmt = (
                select(func.count(Memory.id))
                .where(Memory.agent_id == agent_id)
                .where(~Memory.memory_type.startswith("archived_"))
                .where(Memory.memory_type != "conversation_summary")
            )
            count_result = await session.execute(count_stmt)
            count = count_result.scalar() or 0

            # Estimate total tokens (sum content lengths / 4)
            size_stmt = (
                select(func.sum(func.length(Memory.content)))
                .where(Memory.agent_id == agent_id)
                .where(~Memory.memory_type.startswith("archived_"))
                .where(Memory.memory_type != "conversation_summary")
            )
            size_result = await session.execute(size_stmt)
            total_chars = size_result.scalar() or 0
            estimated_tokens = total_chars / 4

        count_threshold = int(50 * self._compaction_threshold)
        token_threshold = int(8000 * self._compaction_threshold)

        return count > count_threshold or estimated_tokens > token_threshold

File: internal/memory/compaction.go
package memory

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"github.com/godinj/drem-orchestrator/internal/model"
)

// OrchestratorSnapshot represents a checkpoint of orchestrator state.
type OrchestratorSnapshot struct {
	ActiveTasks    []uuid.UUID `json:"active_tasks"`
	PendingReviews []uuid.UUID `json:"pending_reviews"`
	ActiveAgents   []uuid.UUID `json:"active_agents"`
	StaleAgents    []uuid.UUID `json:"stale_agents"`
	LastCheckpoint time.Time   `json:"last_checkpoint"`
}

// SaveOrchestratorState saves a checkpoint of the current orchestrator state
// as a Memory record with type "orchestrator_state".
//
//  1. Query active tasks (BACKLOG, PLANNING, IN_PROGRESS, MERGING)
//  2. Query pending reviews (PLAN_REVIEW, MANUAL_TESTING)
//  3. Query active agents (WORKING status)
//  4. Store as a Memory with metadata containing the snapshot
func SaveOrchestratorState(db *gorm.DB, orchestratorAgentID uuid.UUID) error {
	now := time.Now()

	// 1. Active tasks
	var activeTasks []model.Task
	err := db.Where("status IN ?", []string{
		string(model.StatusBacklog),
		string(model.StatusPlanning),
		string(model.StatusInProgress),
		string(model.StatusMerging),
	}).Find(&activeTasks).Error
	if err != nil {
		return fmt.Errorf("save orchestrator state: query active tasks: %w", err)
	}

	activeTaskIDs := make([]uuid.UUID, len(activeTasks))
	for i, t := range activeTasks {
		activeTaskIDs[i] = t.ID
	}

	// 2. Pending reviews
	var pendingReviews []model.Task
	err = db.Where("status IN ?", []string{
		string(model.StatusPlanReview),
		string(model.StatusManualTesting),
	}).Find(&pendingReviews).Error
	if err != nil {
		return fmt.Errorf("save orchestrator state: query pending reviews: %w", err)
	}

	pendingReviewIDs := make([]uuid.UUID, len(pendingReviews))
	for i, t := range pendingReviews {
		pendingReviewIDs[i] = t.ID
	}

	// 3. Active agents
	var activeAgents []model.Agent
	err = db.Where("status = ?", string(model.AgentWorking)).Find(&activeAgents).Error
	if err != nil {
		return fmt.Errorf("save orchestrator state: query active agents: %w", err)
	}

	activeAgentIDs := make([]uuid.UUID, len(activeAgents))
	for i, a := range activeAgents {
		activeAgentIDs[i] = a.ID
	}

	snapshot := OrchestratorSnapshot{
		ActiveTasks:    activeTaskIDs,
		PendingReviews: pendingReviewIDs,
		ActiveAgents:   activeAgentIDs,
		StaleAgents:    nil,
		LastCheckpoint: now,
	}

	// Marshal snapshot to metadata
	snapshotBytes, err := json.Marshal(snapshot)
	if err != nil {
		return fmt.Errorf("save orchestrator state: marshal snapshot: %w", err)
	}

	var metadata model.JSONField
	if err := json.Unmarshal(snapshotBytes, &metadata); err != nil {
		return fmt.Errorf("save orchestrator state: unmarshal to metadata: %w", err)
	}

	// 4. Store as Memory
	mem := &model.Memory{
		ID:         uuid.New(),
		AgentID:    orchestratorAgentID,
		Content:    fmt.Sprintf("Orchestrator checkpoint at %s", now.Format(time.RFC3339)),
		MemoryType: "orchestrator_state",
		Metadata:   metadata,
		CreatedAt:  now,
	}

	if err := db.Create(mem).Error; err != nil {
		return fmt.Errorf("save orchestrator state: create memory: %w", err)
	}

	return nil
}

// RestoreOrchestratorState loads the last checkpoint and reconciles stale
// agents.
//
//  1. Load latest "orchestrator_state" memory for the given orchestrator agent
//  2. Parse snapshot from metadata
//  3. Check heartbeats for active agents — mark stale ones
func RestoreOrchestratorState(db *gorm.DB, orchestratorAgentID uuid.UUID) (*OrchestratorSnapshot, error) {
	// 1. Load latest orchestrator_state memory
	var mem model.Memory
	err := db.
		Where("agent_id = ?", orchestratorAgentID).
		Where("memory_type = ?", "orchestrator_state").
		Order("created_at DESC").
		First(&mem).Error
	if err != nil {
		return nil, fmt.Errorf("restore orchestrator state: load checkpoint: %w", err)
	}

	// 2. Parse snapshot from metadata
	metadataBytes, err := json.Marshal(mem.Metadata)
	if err != nil {
		return nil, fmt.Errorf("restore orchestrator state: marshal metadata: %w", err)
	}

	var snapshot OrchestratorSnapshot
	if err := json.Unmarshal(metadataBytes, &snapshot); err != nil {
		return nil, fmt.Errorf("restore orchestrator state: parse snapshot: %w", err)
	}

	// 3. Check heartbeats for active agents — identify stale ones
	staleThreshold := time.Now().Add(-5 * time.Minute)
	var staleAgentIDs []uuid.UUID

	for _, agentID := range snapshot.ActiveAgents {
		var agent model.Agent
		if err := db.Where("id = ?", agentID).First(&agent).Error; err != nil {
			// Agent no longer exists — consider it stale
			staleAgentIDs = append(staleAgentIDs, agentID)
			continue
		}

		// Check heartbeat
		if agent.HeartbeatAt == nil || agent.HeartbeatAt.Before(staleThreshold) {
			staleAgentIDs = append(staleAgentIDs, agentID)
			// Mark agent as dead
			db.Model(&model.Agent{}).Where("id = ?", agentID).Update("status", string(model.AgentDead))
		}
	}

	snapshot.StaleAgents = staleAgentIDs

	return &snapshot, nil
}

// ShouldCompact returns true if an agent's memory count exceeds thresholds.
// Threshold: memory_count > 50 * compactionThreshold OR estimated tokens >
// 8000 * compactionThreshold.
func ShouldCompact(db *gorm.DB, agentID uuid.UUID, compactionThreshold float64) (bool, error) {
	// Count non-archived memories
	var count int64
	err := db.Model(&model.Memory{}).
		Where("agent_id = ?", agentID).
		Where("memory_type NOT LIKE ?", "archived_%").
		Where("memory_type != ?", "conversation_summary").
		Count(&count).Error
	if err != nil {
		return false, fmt.Errorf("should compact: count memories: %w", err)
	}

	if float64(count) > 50*compactionThreshold {
		return true, nil
	}

	// Estimate tokens as sum(len(content)) / 4
	var totalLength struct {
		Total int64
	}
	err = db.Model(&model.Memory{}).
		Select("COALESCE(SUM(LENGTH(content)), 0) as total").
		Where("agent_id = ?", agentID).
		Where("memory_type NOT LIKE ?", "archived_%").
		Where("memory_type != ?", "conversation_summary").
		Scan(&totalLength).Error
	if err != nil {
		return false, fmt.Errorf("should compact: estimate tokens: %w", err)
	}

	estimatedTokens := float64(totalLength.Total) / 4.0
	if estimatedTokens > 8000*compactionThreshold {
		return true, nil
	}

	return false, nil
}
